src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                _csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
 
-                ThrowTerminatingError(errorRecord);
-                parseOptions = null;
+                if (UsingNamespace.Length > 0)
+                {
+                    _csharpCompilationOptions = _csharpCompilationOptions.WithUsings(UsingNamespace);
+                }
             }
 
-            SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(source, parseOptions);
-            var references = s_defaultAssemblies.Value;
-            if (ReferencedAssemblies.Length > 0)
+            if (String.Equals(ParameterSetName, "FromPath", StringComparison.OrdinalIgnoreCase) ||
+                String.Equals(ParameterSetName, "FromLiteralPath", StringComparison.OrdinalIgnoreCase))
             {
-                var tempReferences = new List<PortableExecutableReference>(s_autoReferencedAssemblies.Value);
-                foreach (string assembly in ReferencedAssemblies)
+                foreach (string file in _paths)
                 {
-                    if (string.IsNullOrWhiteSpace(assembly)) { continue; }
-                    string resolvedAssemblyPath = ResolveAssemblyName(assembly, true);
+                    using (var sourceFile = new FileStream(file, FileMode.Open))
+                    {
+                        var sourceText = SourceText.From(sourceFile);
+                        _syntaxTrees.Add(CSharpSyntaxTree.ParseText(sourceText, _csharpParseOptions, file));
+                    }
+                }
+            }
+            else if (String.Equals(ParameterSetName, "FromMember", StringComparison.OrdinalIgnoreCase))
+            {
+                _sourceCode = GenerateTypeSource(Namespace, Name, _sourceCode, Language);
 
-                    // Ignore some specified reference assemblies
-                    string fileName = PathType.GetFileName(resolvedAssemblyPath);
-                    if (s_refAssemblyNamesToIgnore.Value.Contains(fileName))
+                var sourceText = SourceText.From(_sourceCode);
+                _syntaxTrees.Add(CSharpSyntaxTree.ParseText(sourceText, _csharpParseOptions));
+            }
+            else if (String.Equals(ParameterSetName, "FromSource", StringComparison.OrdinalIgnoreCase))
+            {
+                var sourceText = SourceText.From(_sourceCode);
+                _syntaxTrees.Add(CSharpSyntaxTree.ParseText(sourceText, _csharpParseOptions));
+            }
+            else
+            {
+                Diagnostics.Assert(false, "Invalid parameter set: {0}", this.ParameterSetName);
+            }
+
+            if (CompileOnly)
+            {
+                CSharpCompileToAssembly();
+            }
+            else
+            {
+                // if the source code was already compiled and loaded and not changed
+                // we get the assembly from the cache.
+                if (isSourceCodeUpdated(out Assembly assembly))
```

Comment:
I think it would be better if you make `_syntaxTree` also a local variable. It can be passed to `IsSourceCodeUpdated` and `CSharpCompileToAssembly`.
I think it's better to make it clear what a method consumes and produces in its signature, and having the arguments like `compilationOptions, emitOptions, syntaxTree` passed into methods helps it.

Reply:
I am waiting https://github.com/PowerShell/PowerShell/pull/6141#discussion_r171310625

Reply:
Fixed.